gr-digital/lib/meas_evm_cc_impl.cc

```diff
@@ -0,0 +1,78 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2020 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "meas_evm_cc_impl.h"
+#include <gnuradio/io_signature.h>
+
+namespace gr {
+namespace digital {
+
+meas_evm_cc::sptr meas_evm_cc::make(constellation_sptr cons, evm_measurement_t meas_type)
+{
+    return gnuradio::get_initial_sptr(new meas_evm_cc_impl(cons, meas_type));
+}
+
+meas_evm_cc_impl::meas_evm_cc_impl(constellation_sptr cons, evm_measurement_t meas_type)
+    : gr::sync_block("meas_evm_cc",
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)),
+                     gr::io_signature::make(1, 1, sizeof(float))),
+      d_cons(cons),
+      d_meas_type(meas_type)
+{
+
+    d_cons_points = d_cons->points();
+
+    double sum = 0.0;
+    int N = d_cons_points.size();
+    for (int i = 0; i < N; i++) {
+        sum += pow(real(d_cons_points[i]), 2.0) + pow(imag(d_cons_points[i]), 2.0);
+    }
+    sum /= (double)N;
+    d_cons_mag_sq = sum;
+    d_cons_mag = sqrt(sum);
+}
+
+double meas_evm_cc_impl::mag(gr_complex x)
+{
+    return sqrt(pow(real(x), 2.0) + pow(imag(x), 2.0));
+}
+
+meas_evm_cc_impl::~meas_evm_cc_impl() {}
+
+int meas_evm_cc_impl::work(int noutput_items,
+                           gr_vector_const_void_star& input_items,
+                           gr_vector_void_star& output_items)
+{
+    const gr_complex* in = (const gr_complex*)input_items[0];
+    float* out = (float*)output_items[0];
+
+    // Compare incoming symbols to the constellation decision points
+    for (int s = 0; s < noutput_items; s++) {
+        gr_complex u_n = in[s];
+        gr_complex decision;
+        d_cons->map_to_points(d_cons->decision_maker(&u_n), &decision);
+        double err_mag = std::abs(decision - u_n);
```

Comment:
why the conversion from single to double precision here?

Reply:
most likely sloppiness